mithril-aggregator/src/services/cardano_transactions_importer.rs

```diff
@@ -0,0 +1,328 @@
+use async_trait::async_trait;
+use mithril_common::cardano_transaction_parser::TransactionParser;
+use mithril_common::entities::{CardanoTransaction, ImmutableFileNumber};
+use mithril_common::signable_builder::TransactionsImporter;
+use mithril_common::StdResult;
+use slog::{debug, Logger};
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+#[cfg(test)]
+use mockall::automock;
+
+/// Cardano transactions store
+#[cfg_attr(test, automock)]
+#[async_trait]
+pub trait TransactionStore: Send + Sync {
+    /// Get the highest known transaction beacon
+    async fn get_highest_beacon(&self) -> StdResult<Option<ImmutableFileNumber>>;
+
+    /// Get stored transactions up to the given beacon
+    async fn get_up_to(
+        &self,
+        immutable_file_number: ImmutableFileNumber,
+    ) -> StdResult<Vec<CardanoTransaction>>;
+
+    /// Store list of transactions
+    async fn store_transactions(&self, transactions: Vec<CardanoTransaction>) -> StdResult<()>;
+}
+
+/// Import and store [CardanoTransaction].
+pub struct CardanoTransactionsImporter {
+    transaction_parser: Arc<dyn TransactionParser>,
+    transaction_store: Arc<dyn TransactionStore>,
+    logger: Logger,
+    rescan_offset: Option<usize>,
+    dirpath: PathBuf,
+}
+
+impl CardanoTransactionsImporter {
+    /// Constructor
+    ///
+    /// About `rescan_offset`: if Some(x) the importer will be asked to rescan the previous 'x'
+    /// immutables starting from the highest immutable known in the store.
+    /// This is useful when one of the last immutable was not full scanned.
+    pub fn new(
+        transaction_parser: Arc<dyn TransactionParser>,
+        transaction_store: Arc<dyn TransactionStore>,
+        dirpath: &Path,
+        rescan_offset: Option<usize>,
+        logger: Logger,
+    ) -> Self {
+        Self {
+            transaction_parser,
+            transaction_store,
+            logger,
+            rescan_offset,
+            dirpath: dirpath.to_owned(),
+        }
+    }
+
+    async fn get_starting_beacon(&self) -> StdResult<Option<u64>> {
+        let highest = self.transaction_store.get_highest_beacon().await?;
+        let rescan_offset = self.rescan_offset.unwrap_or(0);
+        let highest = highest.map(|h| (h + 1).saturating_sub(rescan_offset as u64));
+        Ok(highest)
+    }
+
+    async fn parse_and_store_missing_transactions(
+        &self,
+        from: Option<u64>,
+        until: ImmutableFileNumber,
```

Comment:
IMO, it would have been easier to use a `Range<ImmutableFileNumber>` here.
Also it should be `from: Option<ImmutableFileNumber>`?